### PR TITLE
BLE: announce heart rate service in ADV packet.

### LINKED ISSUE #895: Expose the HeartRate service UUID in BLE ADV packet
### Verification

- [X] I searched for similar feature request and found none was relevant.

### Pitch us your idea!

Expose the HR service UUID in ADV so that 3rd party application detects the PineTime as a HR device

### Description

OpenTracks reported that a PineTime running InfiniTime is not detected as a HR device because the UUID of the HR service is not exposed in the ADV packet.

See https://github.com/OpenTracksApp/OpenTracks/issues/1063 and https://github.com/OpenTracksApp/OpenTracks/pull/1069.

Adding this UUID in ADV would allow companion apps to detect the PineTime as a HR device without needed to connect and discover the services running on the PineTime.

Basically, we would need to add the specific UUID here : https://github.com/InfiniTimeOrg/InfiniTime/blob/develop/src/components/ble/NimbleController.cpp#L162

- [ ] What are the implications of adding multiple UUIDs to the ADV packet?

### PR DESCRIPTION
**Pull request by dennisguse**
Fixes #895.

Not tested due to lack of device.
Did not test compiling locally :)

### CODE REVIEW

File: src/components/ble/NimbleController.cpp [hunk 1]
```diff
@@ -158,9 +158,11 @@ void NimbleController::StartAdvertising() {
     adv_params.itvl_max = 1651;
   }
 
+  ble_uuid128_t serviceUUIDs[] = {dfuServiceUuid, heartRateServiceUuid};
+
   fields.flags = BLE_HS_ADV_F_DISC_GEN | BLE_HS_ADV_F_BREDR_UNSUP;
-  fields.uuids128 = &dfuServiceUuid;
-  fields.num_uuids128 = 1;
+  fields.uuids128 = &serviceUUIDs;
```
Comment (yehoshuapw) [inline src/components/ble/NimbleController.cpp:164]: I don't think you need the "&", now that it's really a array.

Comment (dennisguse) [reply src/components/ble/NimbleController.cpp:164]: Thx :)

**Comment by yehoshuapw** [comment]
I can't tell if it's doing anything different - but I built and pushed this change, and nothing is broken.

**Comment by trman** [comment]
i have tested and don't resolve the initial issue : 
opentrack 3.23 don't seen this pr infinitime as hr device

**Comment by dennisguse** [comment]
@trman Thanks for testing!
The UUID should be `0000180d-0000-1000-8000-00805f9b34fb` and I hope I encoded it correctly.
I came up with `0xFB, 0x34, 0x9B, 0x5F, 0x80, 0x00, 0x00, 0x80, 0x00, 0x10, 0x00, 0x00, 0x0D, 0x18, 0x00, 0x00`.

What I did:
```
package de.dennisguse.opentracks.util;

import static org.junit.Assert.assertEquals;

import androidx.annotation.NonNull;
import androidx.test.ext.junit.runners.AndroidJUnit4;

import org.junit.Test;
import org.junit.runner.RunWith;

import java.nio.ByteBuffer;
import java.util.UUID;

@RunWith(AndroidJUnit4.class)
public class UUIDUtilsTest {

    @Test
    public void test() {

        byte[] dfuUUID = new byte[]{0x23, (byte) 0xD1, (byte) 0xBC, (byte) 0xEA, 0x5F, 0x78, 0x23, 0x15, (byte) 0xDE, (byte) 0xEF, 0x12, 0x12, 0x30, 0x15, 0x00, 0x00};
        assertEquals("00001530-1212-efde-1523-785feabcd123", fromBytes(reverse(dfuUUID)).toString());

        assertEquals(toString(new byte[]{(byte) 0xFB, 0x34, (byte) 0x9B, 0x5F, (byte) 0x80, 0x00, 0x00, (byte) 0x80, 0x00, 0x10, 0x00, 0x00, 0x0D, 0x18, 0x00, 0x00}), toString(reverse(toBytes(BluetoothUtils.HEART_RATE_SERVICE_UUID))));
        assertEquals(BluetoothUtils.HEART_RATE_SERVICE_UUID, fromBytes(reverse(new byte[]{(byte) 0xFB, 0x34, (byte) 0x9B, 0x5F, (byte) 0x80, 0x00, 0x00, (byte) 0x80, 0x00, 0x10, 0x00, 0x00, 0x0D, 0x18, 0x00, 0x00})));
    }

    public static String toString(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append("0x" + String.format("%02X ", b));
        }
        return  sb.toString();
    }

    public static byte[] reverse(byte[] bytes) {
        byte[] reversed = new byte[bytes.length];
        for (int i = 0; i < bytes.length; i++) {
            reversed[bytes.length - 1 - i] = bytes[i];
        }
        return reversed;
    }

    public static UUID fromBytes(byte[] bytes) {
        ByteBuffer byteBuffer = ByteBuffer.wrap(bytes);
        long mostSignificant = byteBuffer.getLong();
        long lestSignificant = byteBuffer.getLong();
        return new UUID(mostSignificant, lestSignificant);
    }

    public static byte[] toBytes(@NonNull UUID uuid) {
        ByteBuffer byteBuffer = ByteBuffer.allocate(16);
        byteBuffer.putLong(uuid.getMostSignificantBits());
        byteBuffer.putLong(uuid.getLeastSignificantBits());
        return byteBuffer.array();
    }
}
```

Disclaimer: Again lack of test device and I never used nimble and don't have local setup to compile Infinitime.

**Comment by dennisguse** [comment]
Side note: I don't know if Android parses all service UUIDs from an ADV packet.
Never had any chance to test it :/

**Comment by Avamander** [comment]
It's usually not Android, it's the app - that can then reason about a device. Though instead of advertising more, OpenTracks could check for the service and use it if found.

**Comment by dennisguse** [comment]
@Avamander About a year ago I thought exactly the same, but sadly that's not true.
An explanation can be found here: https://github.com/OpenTracksApp/OpenTracks/issues/1063#issuecomment-1002687480

Relevant code: https://github.com/OpenTracksApp/OpenTracks/blob/55d3502ff91b419b36b9579f89c001e5b206ff99/src/main/java/de/dennisguse/opentracks/settings/bluetooth/BluetoothLeSensorPreference.java#L213

**Comment by Avamander** [comment]
It is true, you can filter beacons for the devices you're looking for and then initiating a service discovery. Alternatively, let the user pick a device and then do it. Costly or not, it doesn't matter in that case.

**Comment by trman** [comment]
so another test  with the nightly of open track (add on toogle in order to not filter uuid) 

case infinitime is securely connected with gadgetbridge:
- uuid filtered : infinitme not listed
- uuid not filterd : infinitime not seen by opentrack

case infinitime is disconnected of gadgetbridge:
- uuid filtered : infinitme not listed
- uuid not filterd : infinitime  seen by opentrack and usable only when infinitme screen is ON (otherwise it's 0)

for what is see , it can show two (three?) things:
firrst the uuid is not accepted with opentrack 
if infinitme is connected securely to gadgetbrige , opentrack can't see infinitme as bluetooth device 
if infitme is connected to opentrack , the hr task is not functional/usable when the screen is off in infinitme and that make infinitme useless as hr device

**Comment by dennisguse** [comment]
@trman That is expected - most devices do not send AVD when connected to some other application/ device.
Could you also test if you only AVD the heart rate service UUID?

PS/ RunnerUp also filters by AVD UUID: https://github.com/jonasoreland/runnerup

**Comment by trman** [comment]
> @trman That is expected - most devices do not send AVD when connected to some other application/ device. Could you also test if you only AVD the heart rate service UUID?
> 
> PS/ RunnerUp also filters by AVD UUID: https://github.com/jonasoreland/runnerup

Hi , @dennisguse , @JF002 

as said by  @dennisguse , RunnerUp also filters by AVD UUID so here is the test with runner up :

case infinitime is securely connected ( or not) with gadgetbridge:
-infinitime  seen by RunnerUp and usable but fc report only when infinitime screen is ON (otherwise it's 0)

case infinitime is disconnected of gadgetbridge:
-infinitime  seen by RunnerUp and usable but fc report only when infinitime screen is ON (otherwise it's 0)

for what is see , it can show two  things:
1) the uuid is  accepted by RunnerUp (unlike opentracks)
2) if infinitime is connected securely to gadgetbridge  , RunnerUp do accept infinitme as hr device on spot 
but hr  is not functional/usable when the screen is off in infinitime and that make infinitime useless as hr device

**Comment by trman** [comment]
> @trman That is expected - most devices do not send AVD when connected to some other application/ device. Could you also test if you only AVD the heart rate service UUID?
> 
> PS/ RunnerUp also filters by AVD UUID: https://github.com/jonasoreland/runnerup

Hi , @dennisguse , @JF002 

Since @dennisguse , said RunnerUp also filters by AVD UUID 

so here is the test with RunnerUp (from fdroid)  :

case infinitime is securely connected with gadgetbridge:
-infinitime  seen by RunnerUp and usable but fc report only when infinitime screen is ON (otherwise it's 0)

case infinitime is disconnected of gadgetbridge:
-infinitime  not seen by RunnerUp  (it don't show on scan device)

for what is see , it can show two  things:
1) the uuid is not  accepted by RunnerUp
2) if infinitime is connected securely to gadgetbridge  , RunnerUp do accept infinitme as hr device on spot 
but hr  is not functional/usable when the screen is off in infinitime and that make infinitime useless as hr device


it's seem that RunnerUp do/may use another way (that complement/replace uuid)  to connect to hr device , (and that way work without the need to disconnect of gadgetbridge) maybe opentracks dev should use a similar way

it may be related to the CompanionDevice API  ( that gadgetbridge use as well) that allow to share the fc to other apps

**Comment by Avamander** [comment]
I guess in the end this PR's behaviour is fine, but I have a few small concerns with how it'll behave in the future.

The PR's behaviour would be nice to test with the encrypted+secured connection flag on the characteristic, because that is on the future roadmap to only allow secure methods of accessing the PT. It doesn't have to be done now, but it seems like it might break.

**Comment by Avamander** [comment]
@trman 
Companion device pairing (CDM) API only sets a specific system database flag that grants the app extra rights (which allows better resistance to OOM kills). It should not affect pairing/bonding, it's not a (pre)requisite.

GB simply does service discovery for each interesting device, attaches itself to what it can understand. It doesn't care too much about only the adv. beacon like the other apps in question.

**Comment by trman** [comment]
well @Avamander   as far this new uuid seem concerned , runnerup is more up to task for the secure connection than opentracks (at least for now ) with the new uuid 

the next version of opentrack may allow to search device device without filtering

**Comment by trman** [comment]
> Companion device pairing (CDM) API only sets a specific system database flag that grants the app extra rights (which allows better resistance to OOM kills). It should not affect pairing/bonding, it's not a (pre)requisite.
@Avamander it seem you are mistaken about the paring : as per this https://developer.android.com/guide/topics/connectivity/companion-device-pairing and https://codeberg.org/Freeyourgadget/Gadgetbridge/wiki/Companion-Device-pairing
this show that the pairing is faster , more stable and easier with companion device.

**Comment by Avamander** [comment]
@trman Well in some ways CDM is "faster" and "more secure" but not in the ways you're thinking or we're talking. Stability wise, definitely not better, sucks major \*\*\*\* with dual-protocol devices.

**Comment by JF002** [comment]
Gentlemen, please try to stay focus on the topic of this PR : add the HR service UUID in the advertising package so companion apps can detect the PineTime as a HR device without having to start the service discovery procedure.

InfiniTime allows only 1 connection (that's why opentracks cannot see the PineTime once it's connected to Gadgetbridge), does the HR acquisition only when the display is ON and do not implement CDM. Those are known behaviors and out of the scope of this PR. Please open new issues or comment the existing ones if you want to talk about those topics.

In my opinion, the only info we need here is to check that the UUID this PR adds to the advertising packet is the one expected by companion apps to detect a HR sensor.

**Comment by trman** [comment]
> In my opinion, the only info we need here is to check that the UUID this PR adds to the advertising packet is the one expected by companion apps to detect a HR sensor.

as far i'am concerned , @JF002  , i have done an exhaustif tests that allow to check if uid is read in opentracks (and as requested runnerup)
in order to not have misunderstanding , i'am testing with gadgetbridge and without because user might not usnstand why it's not working.
 in fact for the unfiltered opentrack , at first i did'nt understand that gadgetbridge was blocking !
 
 so what you think might be 'obvious' , can very well not be ovious to user....

**Comment by trman** [comment]
in order to have a proper test , i will add shortly a test with opentracks and runnerup without this pr (well it's more for runnerup actually ) in order to check if  runnerup do retreive the adv beacon or not

**Comment by trman** [comment]
here is the test WITHOUT this pr :

so here is the test with RunnerUp (from fdroid)  :

case infinitime is securely connected with gadgetbridge:
-infinitime  seen by RunnerUp and usable 

case infinitime is disconnected of gadgetbridge:
-infinitime  is automaticaly connected as hrm device but when searching hrm device oen option int's not found

**Comment by trman** [comment]
so as far runner up is concerned this beacon is not used to connect to infinitme as hr device

**Comment by dennisguse** [comment]
@trman this patch would just enable OpenTracks to search for the device for device selection; once the MAC address is stored, the ADV packages are not used anymore.
Moreover, RunnerUp checks for bonded aka paired devices - OpenTracks doesn't do this for device discovery.
Don't know if OpenTracks should though.

Relevant code from RunnerUp:
`AndroidBLEHRProvider.java`: 
```        if (AVOID_SCAN_WITH_UUID)
            // Android 4.3
            btAdapter.startLeScan(mLeScanCallback);
        else {
            if (mSupportPaired) {
                for (BluetoothDevice btDeviceThis : btAdapter.getBondedDevices()) {
                    if (btDeviceThis.getType() != BluetoothDevice.DEVICE_TYPE_LE) {
                        log("Ignoring paired non BLE device: " + btDeviceThis.getName());
                        continue;
                    }

                    // Paired device, for instance Huami/Amazfit Bip S could be supported
                    log("Trying paired generic BLE device: " + btDeviceThis.getName());
                    mLeScanCallback.onLeScan(btDeviceThis, 0, null);
                }
            }
            btAdapter.startLeScan(SCAN_UUIDS, mLeScanCallback);
        }
```

**Comment by JF002** [comment]
@trman I'll repeat for the last time : we **know** that InfiniTime accepts only 1 connection, we **know** InfiniTime acquires HR data only when the display is ON. This is confusing for you and other users, right. So let's talk about those topics in their respective issues and/or PR. This PR does not cover these points so **please STOP mentionning them in each and every one of your comment**. It only add noise and confusion in this simple "5 lines" PR.

And I also repeat : the **only** thing that matters here is : are companion apps that only parse the ADV packet (like OpenTracks) able to detect InfiniTime as a HR device. Which means that RunnerUp is **not** covered by this PR as it checks for paired device.

**Comment by trman** [comment]
> And I also repeat : the **only** thing that matters here is : are companion apps that only parse the ADV packet (like OpenTracks) able to detect InfiniTime as a HR device. Which means that RunnerUp is **not** covered by this PR as it checks for paired device.

id'nt know this : it was dennisguse that said that runner up was using uuid , otherwise i would not have bothered do a test for it

**Comment by trman** [comment]
for the opentracks : the pr not working for the filtered search

**Comment by dennisguse** [comment]
How can we proceed from here?
I don't if this an issue with the provided UUID (I might gotten something wrong), the Android BLE discovery subsystem, or something else.
Any suggestions?

@trman Could you record an ADV package (e.g.g, using `bluetoothctl`) and check if it looks okay?

PS/ today I ordered a PineTime, but will take up to 6 weeks :shrug:

**Comment by JF002** [comment]
@dennisguse Sorry for the delay, I got busy on other things and on releasing 1.8 :)

So, I did some test, and I have to admit I'm a bit confused: Using your branch, the display would display nothing... I'm not sure why... 

Then I applied your changes on current develop, and InfiniTime ran correctly, but advertising did not seem right. 
Here is a screenshot of NRFConnect showing the advertising of InfiniTime without your changes:
image

And the same screenshot with your changes:
image

As you can see, there UUIDs disappeared completely, which is obviously not what we expected.
I guess we'll have to dig a bit in NimBLE to see how to manage multiple UUIDs in the advertissement...

I also tried by adding only the HR UUID to check that it works correctly with OpenTracks : it works!
NRFConnect show the UUID:
image

And OpenTracks lists my PineTime even when "Only show Bluetooth devices that announce required services" is enabled :+1:

**Comment by dennisguse** [comment]
@JF002 No worries; thanks for testing!

**Comment by JF002** [comment]
I'll try to do more tests when I get the time.
@evergreen22 @geekbozu  any idea about [this](https://github.com/InfiniTimeOrg/InfiniTime/pull/897#issuecomment-1007673011) issue?

**Comment by dennisguse** [comment]
Side not: the heart rate service UUID is already defined in HeartRateService.h and should probably be taken from there.

And I checked the nimble code and the relevant code that prepares the ADV is this function `ble_hs_adv_set_array_uuid128`.

**Comment by dennisguse** [comment]
@JF002 I now got a sealed PineTime and could do some testing.
Should we continue this or postpone this PR?

**Comment by JF002** [comment]
@dennisguse We can of course continue to work on this PR. Unfortunately, I haven't had the opportunity to do more tests since my last comment :/

**Comment by trman** [comment]
> @JF002 I now got a sealed PineTime and could do some testing.

@dennisguse did you have tried , with your sealed pinetme , some test of your code?
if don't , you should  change (is this not already installed)  the bootloader (and recovery) with this https://github.com/InfiniTimeOrg/pinetime-mcuboot-bootloader/releases/tag/1.0.0

it would make recover the dfu much easier if the dfu installed go wrong

**Comment by dennisguse** [comment]
@trman Sadly, no and time is at a premium the next 2-3 month.
I hope I have some time to fix some ugly things in OpenTracks, but I doubt that I will have time to do this properly.
Thanks for the tip - and luckily the new bootloader was already installed :)

**Comment by florisre** [comment]
Anybody willing to pick this up again? I am happy to help in testing!

**Pull request closed without merging**